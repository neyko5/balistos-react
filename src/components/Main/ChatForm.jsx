import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Input from '../../common/Input';
import Icon from '../../common/Icon';
import Button from '../../common/Button';

const Send = styled.div`
  padding: 10px;
  background: #e8e8e8;
  float: left;
  width: 100%;
`

class ChatForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      message: '',
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    if (!this.state.message.trim()) {
      return;
    }
    this.props.sendMessage(this.state.message);
    this.setState({
      message: '',
    });
  }

  render() {
    return (
      <Send>
        <form onSubmit={this.handleSubmit}>
          <Input
            type="text"
            placeholder="Send a message"
            name="message"
            value={this.state.message}
            onChange={this.handleChange}
          />
          <Button type="submit" green right>
            <Icon message /> Chat
          </Button>
        </form>
      </Send>
    );
  }
}

ChatForm.propTypes = {
  sendMessage: PropTypes.func.isRequired,
};

export default ChatForm;
