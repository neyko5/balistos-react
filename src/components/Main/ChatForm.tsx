import React, { FormEvent } from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import Button from '../common/Button';
import Icon from '../common/Icon';
import Input from '../common/Input';
import { AuthUserType } from '../../types';
import { addChatToPlaylist } from '../../services/firestore.service';

const Send = styled.div`
    padding: 10px;
    background: #e8e8e8;
    width: 100%;
`;

interface Props {
    id: string;
    user: AuthUserType;
}

const ChatForm = (props: Props) => {
    const [message, setMessage] = React.useState('');

    function handleChange(event: FormEvent<HTMLFormElement>): void {
        if (event.target instanceof HTMLInputElement) {
            setMessage(event.target.value);
        }
    }

    function handleSubmit(event: FormEvent<HTMLFormElement>): void {
        event.preventDefault();
        if (!message.trim()) {
            return;
        }
        addChatToPlaylist(props.id, message);
        setMessage('');
    }

    return (
        <Send>
            <form onSubmit={handleSubmit}>
                <Input
                    type="text"
                    placeholder="Send a message"
                    name="message"
                    value={message}
                    onChange={handleChange}
                />
                <Button type="submit" color="green" right={true}>
                    <Icon message={true} /> Chat
                </Button>
            </form>
        </Send>
    );
};

export default connect((state: any) => ({
    user: state.firebase.auth,
}))(ChatForm) as React.ComponentType<any>;
