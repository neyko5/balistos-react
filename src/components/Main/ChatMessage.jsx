// @flow

import React from 'react';
import styled from 'styled-components';

const Message = styled.div`
  margin-bottom: 3px;
  color: #3e414c;
  font-size: 12px;
`;
export const Author = styled.span`
  margin-right: 3px;
  font-weight: 700;
  color: ${props => (props.current ? '#9FA600' : '#3e414c')}
`;

const ChatMessage = (props: {
  username: string,
  message: any,
}) => (
  <Message>
    <Author
      current={props.message.user.username === props.username}
      title={new Date(props.message.createdAt).toDateString()}
    >{props.message.user.username}:
    </Author>
    <span>{props.message.message}</span>
  </Message>
);

export default ChatMessage;
