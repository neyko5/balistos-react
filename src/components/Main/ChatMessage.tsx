import React from 'react';
import styled from 'styled-components';
import { ChatMessageType } from '../../types/index';

const Message = styled.div`
    margin-bottom: 3px;
    color: #3e414c;
    font-size: 12px;
`;
export const Author = styled.span`
    margin-right: 3px;
    font-weight: 700;
    color: ${(props: AuthorProps) => (props.current ? '#9FA600' : '#3e414c')};
`;

interface AuthorProps {
    current: boolean;
}

const ChatMessage = (props: { user: any; message: ChatMessageType }) => (
    <Message>
        <Author
            current={props.message.creator.id === props.user.id}
            title={new Date(props.message.created_at).toDateString()}
        >
            {props.message.creator.name}:
        </Author>
        <span>{props.message.message}</span>
    </Message>
);

export default ChatMessage;
