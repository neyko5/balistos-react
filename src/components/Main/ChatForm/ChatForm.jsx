import React from 'react';
import PropTypes from 'prop-types';

import './ChatForm.css';

const ChatForm = (props) => {
  let message;
  const submitForm = (e) => {
    e.preventDefault();
    if (!message.value.trim()) {
      return;
    }
    props.sendMessage(message.value);
    message.value = '';
  };
  return (
    <div className="send">
      <form onSubmit={submitForm}>
        <input
          type="text"
          placeholder="Send a message"
          ref={(node) => { message = node; }}
        />
        <button type="submit" className="button green chat">
          <i className="icon message" /> Chat</button>
      </form>
    </div>
  );
}

ChatForm.propTypes = {
  sendMessage: PropTypes.func.isRequired,
};

export default ChatForm;
