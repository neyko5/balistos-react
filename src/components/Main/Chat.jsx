import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import ChatMessage from './ChatMessage';
import ChatForm from './ChatForm';
import Icon from '../common/Icon';

const Chatbox = styled.div`
  padding: 12px 10px;
  width: 100%;
  max-height: 300px;
  overflow-y: scroll;
`;

const Title = styled.div`
  font-size: 14px;
  font-weight: 600;
  color: #333333;
  line-height: 26px;
`

const BoxHeader = styled.div`
  width: 100%;
  padding-left: 5px;
  margin-bottom: 5px;
`

const Chat = props => (
  <div>
    <BoxHeader>
      <Icon chat />
      <Title>Chat with your buddies</Title>
    </BoxHeader>
    <Chatbox id="chatbox">
      {props.messages.sort((a, b) => (a.createdAt < b.createdAt ? -1 : 1))
        .map(message => (<ChatMessage
          message={message}
          username={props.username}
          key={message.id}
        />))
      }
    </Chatbox>
    {props.username ? <ChatForm sendMessage={props.sendMessage} /> : undefined}
  </div>
);

Chat.propTypes = {
  username: PropTypes.string,
  sendMessage: PropTypes.func.isRequired,
  messages: PropTypes.arrayOf(
    PropTypes.shape({
      createdAt: PropTypes.string.isRequired,
      message: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
    }),
  ).isRequired,
};


Chat.defaultProps = {
  username: undefined,
};

export default Chat;
