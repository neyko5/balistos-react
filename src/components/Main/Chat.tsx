import React from 'react';
import styled from 'styled-components';

import { ChatMessageType } from '../../types/index';
import Icon from '../common/Icon';
import ChatForm from './ChatForm';
import ChatMessage from './ChatMessage';

const Chatbox = styled.div`
    padding: 12px 10px;
    width: 100%;
    max-height: 300px;
    overflow-y: scroll;
`;

const Title = styled.div`
    font-size: 14px;
    font-weight: 600;
    color: #333333;
    line-height: 26px;
`;

const BoxHeader = styled.div`
    width: 100%;
    padding-left: 5px;
    margin-bottom: 5px;
`;

const Chat = (props: {
    user: any;
    id: string;
    messages: ChatMessageType[];
}) => {
    return (
        <>
            <BoxHeader>
                <Icon chat={true} />
                <Title>Chat with your buddies</Title>
            </BoxHeader>
            <Chatbox id="chatbox">
                {props.messages.map((message) => (
                    <ChatMessage
                        message={message}
                        user={props.user}
                        key={message.id}
                    />
                ))}
            </Chatbox>
            {props.user && !props.user.isEmpty && <ChatForm id={props.id} />}
        </>
    );
};

export default Chat;
