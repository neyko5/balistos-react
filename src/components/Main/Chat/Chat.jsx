import React from 'react';
import PropTypes from 'prop-types';
import ChatMessage from '../ChatMessage';
import ChatForm from '../ChatForm';

import './Chat.css';

const Chat = props => (
  <div className="box">
    <div className="header">
      <i className="icon chat" />
      <div className="title">Chat with your buddies</div>
    </div>
    <div className="chatbox" id="chatbox">
      {props.messages.sort((a, b) => (a.createdAt < b.createdAt ? -1 : 1))
        .map(message => (<ChatMessage
          message={message}
          username={props.username}
          key={message.id}
        />))
      }
    </div>
    {props.username ? <ChatForm sendMessage={props.sendMessage} /> : undefined}
  </div>
);

Chat.propTypes = {
  username: PropTypes.string,
  sendMessage: PropTypes.func.isRequired,
  messages: PropTypes.arrayOf(
    PropTypes.shape({
      createdAt: PropTypes.string.isRequired,
      message: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
    }),
  ).isRequired,
};


Chat.defaultProps = {
  username: undefined,
};

export default Chat;
