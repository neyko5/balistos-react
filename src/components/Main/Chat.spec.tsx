import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import renderer from 'react-test-renderer';

import Chat from './Chat';
import ChatForm from './ChatForm';
import ChatMessage from './ChatMessage';

Enzyme.configure({ adapter: new Adapter() });

const messages = [
    {
        id: '1',
        creator: { id: '2', name: 'Tester 2' },
        message: 'text 1',
        created_at: '2017-04-11T18:18:36.000Z',
    },
    {
        id: '2',
        creator: { id: '1', name: 'Tester 1' },
        message: 'text 2',
        created_at: '2017-05-11T18:18:36.000Z',
    },
    {
        id: '3',
        creator: { id: '2', name: 'Tester 2' },
        message: 'text 3',
        created_at: '2017-06-11T18:18:36.000Z',
    },
];
const user = { id: '1', name: 'Tester 1' };
const noUser = { isEmpty: true };

it('should show chat form for logged in user', () => {
    const chat = shallow(<Chat messages={messages} user={user} id="1" />);
    expect(chat.find(ChatForm).length).toBe(1);
});

it('should show not show chat form for non-logged in user', () => {
    const chat = shallow(<Chat messages={messages} user={noUser} id="1" />);
    expect(chat.find(ChatForm).length).toBe(0);
});

it('should render three messages', () => {
    const chat = shallow(<Chat messages={messages} user={user} id="1" />);
    expect(chat.find(ChatMessage).length).toBe(3);
});

test('should render correctly', () => {
    const chat = renderer.create(
        <Chat messages={messages} user={user} id="1" />
    );
    const tree = chat.toJSON();
    expect(tree).toMatchSnapshot();
});
