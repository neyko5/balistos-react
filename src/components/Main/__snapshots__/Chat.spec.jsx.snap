// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="box"
>
  <div
    className="header"
  >
    <i
      className="icon chat"
    />
    <div
      className="title"
    >
      Chat with your buddies
    </div>
  </div>
  <div
    className="sc-htpNat bfPJqp"
    id="chatbox"
  >
    <div
      className="sc-bdVaJa eNFPzl"
    >
      <span
        className="sc-bwzfXH BCqOY"
        title="Tue Apr 11 2017"
      >
        otheruser
        :
      </span>
      <span>
        text 1
      </span>
    </div>
    <div
      className="sc-bdVaJa eNFPzl"
    >
      <span
        className="sc-bwzfXH BCqOY"
        title="Thu May 11 2017"
      >
        otheruser
        :
      </span>
      <span>
        text 2
      </span>
    </div>
    <div
      className="sc-bdVaJa eNFPzl"
    >
      <span
        className="sc-bwzfXH BCqOY"
        title="Sun Jun 11 2017"
      >
        otheruser
        :
      </span>
      <span>
        text 3
      </span>
    </div>
  </div>
  <div
    className="send"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        name="message"
        onChange={[Function]}
        placeholder="Send a message"
        type="text"
        value=""
      />
      <button
        className="button green chat"
        type="submit"
      >
        <i
          className="icon message"
        />
         Chat
      </button>
    </form>
  </div>
</div>
`;
